runtime/src/main/java/org/corfudb/runtime/collections/ImmutableCorfuTable.java

```diff
@@ -0,0 +1,296 @@
+package org.corfudb.runtime.collections;
+
+import com.google.common.reflect.TypeToken;
+import io.vavr.collection.HashMap;
+import io.vavr.collection.HashSet;
+import io.vavr.collection.Map;
+import io.vavr.collection.Set;
+import io.vavr.control.Option;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.object.ICorfuExecutionContext;
+import org.corfudb.runtime.object.ICorfuSMR;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.stream.Stream;
+
+@Slf4j
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+// TODO: don't need to implement ICorfuSMR?
+public class ImmutableCorfuTable<K, V> implements ICorfuSMR<ImmutableCorfuTable<K, V>> {
+
+    // The "main" map which contains the primary key-value mappings.
+    private Map<K, V> mainMap;
+
+    // Data structures for secondary indexes.
+    private SecondaryIndexesWrapper<K, V> secondaryIndexesWrapper;
+
+    public static <K, V> TypeToken<ImmutableCorfuTable<K, V>> getTableType() {
+        return new TypeToken<ImmutableCorfuTable<K, V>>() {};
+    }
+
+    public ImmutableCorfuTable() {
+        this.mainMap = HashMap.empty();
+        this.secondaryIndexesWrapper = new SecondaryIndexesWrapper<>();
+    }
+
+    public ImmutableCorfuTable(@Nonnull final Index.Registry<K, V> indices) {
+        this.mainMap = HashMap.empty();
+        this.secondaryIndexesWrapper = new SecondaryIndexesWrapper<>(indices);
+    }
+
+    @Override
+    public ImmutableCorfuTable<K, V> getContext(ICorfuExecutionContext.Context context) {
+        return this;
+    }
+
+    /**
+     *
+     * @param key
+     * @return
+     */
+    public V get(@Nonnull K key) {
+        return mainMap.get(key).getOrNull();
+    }
+
+    /**
+     *
+     * @param key
+     * @param value
+     * @return
+     */
+    public ImmutableCorfuTable<K, V> put(@Nonnull K key, @Nonnull V value) {
+        SecondaryIndexesWrapper<K, V> newSecondaryIndexesWrapper = secondaryIndexesWrapper;
+        if (!secondaryIndexesWrapper.isEmpty()) {
+            V prev = get(key);
+            if (prev != null) {
+                newSecondaryIndexesWrapper = newSecondaryIndexesWrapper.unmapSecondaryIndexes(key, prev);
```

Comment:
Doesn't `mapSecondaryIndexes` overwrite the mapping if it existed? It's not clear to me why an `unmapSecondaryIndexes` is needed here.